charms/istio-pilot/src/istioctl.py

```diff
@@ -0,0 +1,214 @@
+
+        Args:
+            binary_file (str): Path to the istioctl binary to be used
+        """
+        self._istioctl_path = istioctl_path
+        self._namespace = namespace
+        self._profile = profile
+
+    @property
+    def _istioctl_flags(self):
+        return [
+            "-s",
+            f"profile={self._profile}",
+            "-s",
+            f"values.global.istioNamespace={self._namespace}",
+        ]
+
+    def install(self):
+        """Wrapper for the `istioctl install` command."""
+        try:
+            subprocess.check_call([self._istioctl_path, "install", "-y", *self._istioctl_flags])
+        except subprocess.CalledProcessError as cpe:
+            error_msg = f"Failed to install istio using istioctl.  Exit code: {cpe.returncode}."
+            logging.error(error_msg)
+            logging.error(f"stdout: {cpe.stdout}")
+            logging.error(f"stderr: {cpe.stderr}")
+
+            raise InstallFailedError(error_msg) from cpe
+
+    def manifest(self) -> str:
+        """Wrapper for the `istioctl manifest generate` command.
+
+        Returns:
+            (str) a YAML string of the Kubernetes manifest for Istio
+        """
+        try:
+            manifests = subprocess.check_output(
+                [self._istioctl_path, "manifest", "generate", *self._istioctl_flags]
+            )
+        except subprocess.CalledProcessError as cpe:
+            error_msg = (
+                f"Failed to generate manifests for istio using istioctl. "
+                f"Exit code: {cpe.returncode}."
+            )
+            logging.error(error_msg)
+            logging.error(f"stdout: {cpe.stdout}")
+            logging.error(f"stderr: {cpe.stderr}")
+
+            raise ManifestFailedError(error_msg) from cpe
+
+        return manifests
+
+    def precheck(self):
+        """Executes `istioctl x precheck` to validate whether the environment can be updated.
+
+        Raises:
+            PrecheckFailedError: if the precheck command fails.
+        """
+        try:
+            subprocess.check_call(
+                [
+                    self._istioctl_path,
+                    "x",
+                    "precheck",
+                ]
+            )
+        except subprocess.CalledProcessError as cpe:
+            raise PrecheckFailedError(
+                "Upgrade failed during `istio precheck` with error code" f" {cpe.returncode}"
+            ) from cpe
+
+    def remove(self):
+        """Removes the Istio installation using istioctl and Lightkube.
+
+        TODO: Should we use `istioctl x uninstall` here instead of lightkube?  It is an
+        experimental feature but included in all istioctl versions we support.
+        """
+        manifest = self.manifest()
+
+        # Render YAML into Lightkube Objects
+        k8s_objects = codecs.load_all_yaml(manifest, create_resources_for_crds=True)
+
+        client = Client()
+        delete_many(client=client, objs=k8s_objects)
```

Comment:
Do we want exception handling around this?
I.e. if status is not 404, we log/raise error.

Reply:
Agreed this is not good.  Since the charm already had a removal procedure, I've just copied it here.  You're right though that it'll fail.  The failure you posted in the main conversation is also a common thing (this charm rarely, if ever, removes itself correctly).  

Since we already had a remove procedure I didn't want to creep the scope of this task to change it, but I think we should fix the remove procedure too. There's an `istioctl x uninstall --purge` that will probably work better for us, and is more in line with how we integrate with everything else. 

What do you think about us opening this as a separate issue, but trying to resolve it right away?

Reply:
ahh damn.  Just tested and `istioctl x uninstall --purge` will also kill `istio-gateway`'s deployment, breaking that charm as well.  So it is not a good candidate for the remove hook here.

Reply:
Sounds good. Let's open separate remove handling issue. There will be more things in remove handler for sure.